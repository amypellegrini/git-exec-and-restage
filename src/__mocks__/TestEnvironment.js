import mockGit from "mock-git";
import mockBin from "mock-bin";
import tmp from "tmp-promise";
import path from "path";
import fs from "mz/fs";
import pmock from "pmock";

export default class TestEnvironment {
  _unmock = [];
  _spyLogFile = undefined;
  __makeSpyJs = undefined;

  async setAllStagedFiles(files = [], failOnHead = false) {
    this._unmock.push(
      await mockGit(
        `${this._makeSpyJs("git")} 
          ${failOnHead ? `
            if (process.argv.indexOf("HEAD") > 1) {
              console.error("fatal: bad revision 'HEAD'");
              process.exitCode = 1;
              return;
            }
          ` : ""}
        for (const file of ${JSON.stringify(files)}) console.log(file);`,
        "diff-index"
      )
    );
  }

  async setPartialFiles(files = []) {
    this._unmock.push(
      await mockGit(
        `${this._makeSpyJs("git")} for (const file of ${JSON.stringify(files)}) console.log(file);`,
        "diff"
      )
    );
  }

  get log() {
    return (async () => {
      if (await fs.exists(this._spyLogFile)) {
        const logContent = await fs.readFile(this._spyLogFile, "utf8");
        return logContent
          .split("\n")
          .filter(Boolean)
          .map(line => JSON.parse(line).join(" "));
      }
      return [];
    })();
  }

  async mockBin(name) {
    this._unmock.push(await mockBin(name, "node", this._makeSpyJs(name)));
  }

  async mockBinError(name) {
    this._unmock.push(await mockBin(name, "false"));
  }

  _makeSpyJs(...prefix) {
    return `
      require("fs").appendFileSync(${JSON.stringify(this._spyLogFile)}, JSON.stringify(${JSON.stringify(prefix)}.concat(process.argv.slice(2))) + "\\n", "utf8");
    `;
  }

  async setup() {
    this._spyLogFile = await tmp.tmpName();
    this._unmock.push(await mockGit(this._makeSpyJs("git"), "add"));
    const cwdMock = pmock.cwd("/path/to/repo");
    this._unmock.push(() => cwdMock.reset());
    await this.setAllStagedFiles();
    await this.setPartialFiles();
  }
  cleanup() {
    for (const unmockOne of this._unmock) {
      unmockOne();
    }
    this._unmock = [];
  }
}
