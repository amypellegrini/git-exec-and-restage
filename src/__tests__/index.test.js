import gitExecAndRestage from "../";
import TestEnvironment from "TestEnvironment";
import path from "path";
import flip from "invert-promise";

describe("git-exec-and-restage", () => {
  const env = new TestEnvironment();
  beforeEach(() => env.setup());
  afterEach(() => env.cleanup());
  it("no command, no args and no files staged", async () => {
    await expect(gitExecAndRestage()).rejects.toMatchSnapshot();
  });
  it("with command, no args and no files", async () => {
    await env.mockBin("prettier");
    await gitExecAndRestage(["prettier"]);
    expect(await env.log).toMatchSnapshot();
  });
  it("throws when command fails", async () => {
    await env.mockBinError("prettier");
    const error = await flip(gitExecAndRestage(["prettier"]));
    expect(error.__isExpectedError).toBe(true);
    expect(error.message).toMatchSnapshot();
  });
  it("with command and arguments, no files", async () => {
    await env.mockBin("prettier");
    await gitExecAndRestage(["prettier", "--write", "--"]);
    expect(await env.log).toMatchSnapshot();
  });
  it("with command, arguments, fully staged file, no partials", async () => {
    await env.mockBin("prettier");
    await env.setAllStagedFiles(["fullystaged.js"]);
    await gitExecAndRestage(["prettier", "--write", "--", "fullystaged.js"]);
    expect(await env.log).toMatchSnapshot();
  });
  it("with command+args, mix of fully and partially staged files", async () => {
    await env.mockBin("prettier");
    await env.setAllStagedFiles(["fullystaged.js", "partiallystaged.js"]);
    await env.setPartialFiles(["partiallystaged.js"]);
    await gitExecAndRestage([
      "prettier",
      "--write",
      "--",
      "fullystaged.js",
      "partiallystaged.js"
    ]);
    expect(await env.log).toMatchSnapshot();
  });
  it("with command+args, files implicit", async () => {
    await env.mockBin("prettier");
    await env.setAllStagedFiles(["fullystaged.js", "partiallystaged.js"]);
    await env.setPartialFiles(["partiallystaged.js"]);
    await gitExecAndRestage(["prettier", "--write", "--"]);
    expect(await env.log).toMatchSnapshot();
  });
  it("with command+args, files implicit with invalid HEAD", async () => {
    await env.mockBin("prettier");
    await env.setAllStagedFiles(
      ["fullystaged.js", "partiallystaged.js"],
      true /* fail on HEAD */
    );
    await env.setPartialFiles(["partiallystaged.js"]);
    await gitExecAndRestage(["prettier", "--write", "--"]);
    expect(await env.log).toMatchSnapshot();
  });
  it("with command+args, pick one of multiple files", async () => {
    await env.mockBin("prettier");
    await env.setAllStagedFiles([
      "fullystaged1.js",
      "fullystaged2.js",
      "partiallystaged1.js",
      "partiallystaged2.js"
    ]);
    await env.setPartialFiles(["partiallystaged1.js", "partiallystaged2.js"]);
    await gitExecAndRestage(["prettier", "--write", "--", "fullystaged1.js"]);
    expect(await env.log).toMatchSnapshot();
  });
  it("with command+args and mixed status, absolute paths", async () => {
    await env.mockBin("prettier");
    await env.setAllStagedFiles(["fullystaged.js", "partiallystaged.js"]);
    await env.setPartialFiles(["partiallystaged.js"]);
    await gitExecAndRestage([
      "prettier",
      "--write",
      "--",
      path.join(process.cwd(), "fullystaged.js"),
      path.join(process.cwd(), "partiallystaged.js")
    ]);
    expect(await env.log).toMatchSnapshot();
  });
});
