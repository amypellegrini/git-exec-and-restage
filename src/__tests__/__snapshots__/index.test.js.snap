// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git-exec-and-restage no command, no args and no files staged 1`] = `[Error: Usage: git-exec-and-restage <command> [...command-args --] [...files]]`;

exports[`git-exec-and-restage throws when command fails 1`] = `"Error running command: Exited with status 1"`;

exports[`git-exec-and-restage with command and arguments, no files 1`] = `
Array [
  "git diff-index --cached --name-only --diff-filter=ACDMRTUXB HEAD --",
  "git diff --name-only --diff-filter=ACDMRTUXB",
  "prettier --write",
]
`;

exports[`git-exec-and-restage with command+args and mixed status, absolute paths 1`] = `
Array [
  "git diff --name-only --diff-filter=ACDMRTUXB fullystaged.js partiallystaged.js",
  "prettier --write /path/to/repo/fullystaged.js /path/to/repo/partiallystaged.js",
  "git add fullystaged.js",
]
`;

exports[`git-exec-and-restage with command+args, files implicit 1`] = `
Array [
  "git diff-index --cached --name-only --diff-filter=ACDMRTUXB HEAD --",
  "git diff --name-only --diff-filter=ACDMRTUXB fullystaged.js partiallystaged.js",
  "prettier --write",
  "git add fullystaged.js",
]
`;

exports[`git-exec-and-restage with command+args, files implicit with invalid HEAD 1`] = `
Array [
  "git diff-index --cached --name-only --diff-filter=ACDMRTUXB HEAD --",
  "git diff-index --cached --name-only --diff-filter=ACDMRTUXB 4b825dc642cb6eb9a060e54bf8d69288fbee4904 --",
  "git diff --name-only --diff-filter=ACDMRTUXB fullystaged.js partiallystaged.js",
  "prettier --write",
  "git add fullystaged.js",
]
`;

exports[`git-exec-and-restage with command+args, mix of fully and partially staged files 1`] = `
Array [
  "git diff --name-only --diff-filter=ACDMRTUXB fullystaged.js partiallystaged.js",
  "prettier --write fullystaged.js partiallystaged.js",
  "git add fullystaged.js",
]
`;

exports[`git-exec-and-restage with command+args, pick one of multiple files 1`] = `
Array [
  "git diff --name-only --diff-filter=ACDMRTUXB fullystaged1.js",
  "prettier --write fullystaged1.js",
  "git add fullystaged1.js",
]
`;

exports[`git-exec-and-restage with command, arguments, fully staged file, no partials 1`] = `
Array [
  "git diff --name-only --diff-filter=ACDMRTUXB fullystaged.js",
  "prettier --write fullystaged.js",
  "git add fullystaged.js",
]
`;

exports[`git-exec-and-restage with command, no args and no files 1`] = `
Array [
  "git diff-index --cached --name-only --diff-filter=ACDMRTUXB HEAD --",
  "git diff --name-only --diff-filter=ACDMRTUXB",
  "prettier",
]
`;
